src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -848,23 +899,123 @@ mergeRuns resolve hfs hbio caching alloc runPaths mergeLevel runs = do
+  credits are atomically subtracted from the PrimVar to get it below the
+  threshold. The number of subtracted credits is then the number of merge steps
+  that will be performed. While doing the merging work, a (more expensive) MVar
+  lock is taken to ensure that the merging work itself is performed only
+  sequentially. If at some point, doing the merge work resulted in the merge
+  being done, then the merge is converted into a new run.
+
+  In the presence of async exceptions, we offer a weaker guarantee regarding
+  consistency of the accumulated, unspent credits and the merge state: a merge
+  /may/ progress more than the number of credits that were taken. If an async
+  exception happens at some point during merging work, then we put back all the
+  credits we took beforehand. This makes the implementation simple, and merges
+  will still finish in time. It would be bad if we did not put back credits,
+  because then a merge might not finish in time, which will mess up the shape of
+  the levels tree.
+
+  The implementation also tracks the total of spent credits, and the number of
+  perfomed merge steps. These are the use cases:
+
+  * The total of spent credits + the total of unspent credits is used by the
+    snapshot feature to restore merge work on snapshot load that was lost during
+    snapshot creation.
+
+  * For simplicity, merges are allowed to do more steps than requested. However,
+    it does mean that once we do more steps next time a batch of work is done,
+    then we should account for the surplus of steps performed by the previous
+    batch. The total of spent credits + the number of performed merge steps is
+    used to compute this surplus, and adjust for it.
+
+    TODO: we should reconsider at some later point in time whether this surplus
+    adjustment is necessary. It does not make a difference for correctness, but it
+    does mean we get a slightly better distribution of work over time. For
+    sensible batch sizes and workloads without many duplicate keys, it probably
+    won't make much of a difference. However, without this calculation the surplus
+    can accumulate over time, so if we're really pedantic about work distribution
+    then this is the way to go
+
+  Async exceptions are allowed to mess up the consistency between the the merge
+  state, the merge steps performed variable, and the spent credits variable.
+  There is an important invariant that we maintain, even in the presence of
+  async exceptions: @merge steps actually performed >= recorded merge steps
+  performed >= recorded spent credits@. TODO: and this makes it correct (?).
+-}
+
+newtype Credit = Credit Int
 
 {-# SPECIALISE supplyCredits ::
-     Credit
+     TableConfig
+  -> Credit
   -> Levels IO h
   -> IO ()
   #-}
+-- | Supply the given amount of credits to each merge in the levels structure.
+-- This /may/ cause some merges to progress.
+--
+-- See Note [Credits]
```

Comment:
This is not ideal in Haddocks, since the note doesn't appear there. Haddock has "named chunks" which we use in a few places (e.g. `$resource-management`), but we also have a few GHC-style notes.

Reply:
It's fine if the piece of documentation is more for internal consumption, as is the case here. But then we should probably make sure it's not referenced in Haddocks.

Reply:
I've removed the reference to the note. Might be good to at some point to describe in the haddocks just at a high level what supplying a credit means, but the note's contents are mostly implementation details so I think it's okay to keep as a note